### ISSUE TITLE
Hoxton.SR4 Kafka Streams and Spring Boot WebFlux 2.2.6 Actuator is not compatible

### ISSUE THREAD

**Issue by davidmelia** [spring-cloud/spring-cloud-stream-binder-kafka#895]
Hi,

Hoxton.SR4 kafka streams and Spring Boot WebFlux 2.2.6 + Actuator is not compatible where you get the error 

```
Caused by: java.lang.IllegalStateException: Unable to map duplicate endpoint operations: [MBean call 'topology'] to topologyEndpoint
	at org.springframework.boot.actuate.endpoint.annotation.EndpointDiscoverer.assertNoDuplicateOperations(EndpointDiscoverer.java:231) ~[spring-boot-actuator-2.2.6.RELEASE.jar:2.2.6.RELEASE]
	at org.springframework.boot.actuate.endpoint.annotation.EndpointDiscoverer.convertToEndpoint(EndpointDiscoverer.java:198) ~[spring-boot-actuator-2.2.6.RELEASE.jar:2.2.6.RELEASE]
```

Downgrading so SR3 is fine.  See an example pom.xml with the problem

```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>
	<parent>
		<groupId>org.springframework.boot</groupId>
		<artifactId>spring-boot-starter-parent</artifactId>
		<version>2.2.6.RELEASE</version>
		<relativePath/> <!-- lookup parent from repository -->
	</parent>
	<groupId>com.melia</groupId>
	<artifactId>temp2</artifactId>
	<version>0.0.1-SNAPSHOT</version>
	<name>temp2</name>
	<description>Demo project for Spring Boot</description>

	<properties>
		<java.version>11</java.version>
		<spring-cloud.version>Hoxton.SR4</spring-cloud.version>
	</properties>

	<dependencies>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-actuator</artifactId>
		</dependency>
            <dependency>
            <groupId>org.springframework.cloud</groupId>
            <artifactId>spring-cloud-stream-binder-kafka-streams</artifactId>
        </dependency>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-webflux</artifactId>
		</dependency>
		<dependency>
			<groupId>org.springframework.cloud</groupId>
			<artifactId>spring-cloud-starter</artifactId>
		</dependency>

	</dependencies>

	<dependencyManagement>
		<dependencies>
			<dependency>
				<groupId>org.springframework.cloud</groupId>
				<artifactId>spring-cloud-dependencies</artifactId>
				<version>${spring-cloud.version}</version>
				<type>pom</type>
				<scope>import</scope>
			</dependency>
		</dependencies>
	</dependencyManagement>

	<build>
		<plugins>
			<plugin>
				<groupId>org.springframework.boot</groupId>
				<artifactId>spring-boot-maven-plugin</artifactId>
			</plugin>
		</plugins>
	</build>

</project>
```

**Comment by sobychacko** [comment]
@davidmelia I am unable to reproduce this issue. I found [similar reporting](https://stackoverflow.com/questions/61712319/error-in-spring-boot-app-unable-to-map-duplicate-endpoint-operations-mbean-c) on Stack Overflow from `Hoxton.SR4` upgrade though. 
Are you seeing this failure on Intellj or any other IDE?
In any case, if you are still having issues, please provide a reproducible standalone application that we can use to triage further. Thanks!

**Comment by sobychacko** [comment]
@davidmelia I just ran into this exact issue with another application on IntelliJ (with `Hoxton.BUILD-SNAPSHOT`) and was able to fix that by disabling the JMX agent in the `Run` profile on the IDE. Please try that as a solution. I am closing this issue now. Feel free to re-open if need be.

**Comment by davidmelia** [comment]
@sobychacko Yes it's an IDE issue.  I'm using the latest Spring STS 4.6.1 and disabling does get around the problem so thanks for that.

However this still seems like a bug as it effects both STS and IntelliJ and was introduced in Hoxton SR4.  What do you think?  Also it's a bit of a pain for developers to have to uncheck this for every new Spring Boot project in their IDE.

**Comment by emenaceb** [comment]
Class TopologyEndpoint added in 3.0.4.RELEASE version doen't meet JMX exporter requirements.
It has an overloaded method "topology" and the exporter seems to validate duplicate operation names.

I think it isn't a IDE related bug but a incompatibility between this endpoint and Spring's JMX exporter.
IDEs enables JMX by default to be able to control spring boot applications, so disabling is not a good idea in most of the cases.


Version Hoxton.SR3 works perfectly as it hasn't the topology actuator endpoint.

Please, @sobychacko could you check that and reopen if that's the problem?

**Comment by sobychacko** [comment]
@davidmelia @emenaceb Sure, I will reopen and look into it.

**Comment by sobychacko** [comment]
@davidmelia @emenaceb The topology endpoint and the methods are renamed. See this [commit](https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/commit/a549810899c67f99447a7e537a3b1e3cbba02973). The same changes are introduced in 3.0.x as well so that the next release, `3.0.6` will contain this change. Please let us know if you encounter further issues with this.

**Comment by jmelby70** [comment]
@sobychacko Even with the fix on 3.0.6 as noted, this is not compatible with spring-boot-actuator 2.3.1. Similar to the original error, I get the following (note MBean call 'kafkaStreamsTopology'). I am using Spring Cloud Hoxton.SR6 and Spring Boot 2.3.1.RELEASE. The duplicate check appears to be counting the overloaded kafkaStreamsTopology method as a duplicate.

```
Caused by: java.lang.IllegalStateException: Unable to map duplicate endpoint operations: [MBean call 'kafkaStreamsTopology'] to topologyEndpoint
	at org.springframework.boot.actuate.endpoint.annotation.EndpointDiscoverer.assertNoDuplicateOperations(EndpointDiscoverer.java:234)
	at org.springframework.boot.actuate.endpoint.annotation.EndpointDiscoverer.convertToEndpoint(EndpointDiscoverer.java:201)
```

**Comment by yonigibbs** [comment]
Just hit the same issue too. Also with the same versions as @jmelby70.

**Comment by kossmoboleat** [comment]
We also still see this error when adding actuator to an app that had `spring-cloud-stream-binder-kafka` and is at "Hoxton.SR7".

**Comment by parroyo** [comment]
We also see this error using `spring-cloud-stream-binder-kafka-streams` and Hoxton.SR4..SR8

**Comment by nekator** [comment]
I can confirm that this bug ist not fixed after renaming the endpoint and the configuration classes.

```
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'jmxMBeanExporter' defined in class path resource [org/springframework/boot/actuate/autoconfigure/endpoint/jmx/JmxEndpointAutoConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.boot.actuate.endpoint.jmx.JmxEndpointExporter]: Factory method 'jmxMBeanExporter' threw exception; nested exception is java.lang.IllegalStateException: Unable to map duplicate endpoint operations: [MBean call 'kafkaStreamsTopology'] to topologyEndpoint
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:655)
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:635)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1336)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1176)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:556)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:516)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:324)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:226)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:322)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:897)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:879)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:551)
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:143)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:758)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:750)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:397)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:315)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1237)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1226)
	at de.protect.garantieprodukte.vertragsanlageprozess.VertragsanlageprozessApplication.main(VertragsanlageprozessApplication.java:16)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.springframework.boot.devtools.restart.RestartLauncher.run(RestartLauncher.java:49)
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.boot.actuate.endpoint.jmx.JmxEndpointExporter]: Factory method 'jmxMBeanExporter' threw exception; nested exception is java.lang.IllegalStateException: Unable to map duplicate endpoint operations: [MBean call 'kafkaStreamsTopology'] to topologyEndpoint
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:185)
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:650)
	... 25 common frames omitted
Caused by: java.lang.IllegalStateException: Unable to map duplicate endpoint operations: [MBean call 'kafkaStreamsTopology'] to topologyEndpoint
	at org.springframework.boot.actuate.endpoint.annotation.EndpointDiscoverer.assertNoDuplicateOperations(EndpointDiscoverer.java:234)
	at org.springframework.boot.actuate.endpoint.annotation.EndpointDiscoverer.convertToEndpoint(EndpointDiscoverer.java:201)
	at org.springframework.boot.actuate.endpoint.annotation.EndpointDiscoverer.convertToEndpoints(EndpointDiscoverer.java:182)
	at org.springframework.boot.actuate.endpoint.annotation.EndpointDiscoverer.discoverEndpoints(EndpointDiscoverer.java:125)
	at org.springframework.boot.actuate.endpoint.annotation.EndpointDiscoverer.getEndpoints(EndpointDiscoverer.java:117)
	at org.springframework.boot.actuate.autoconfigure.endpoint.jmx.JmxEndpointAutoConfiguration.jmxMBeanExporter(JmxEndpointAutoConfiguration.java:95)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154)
	... 26 common frames omitted

```
As a workaround i excluded KafkaStreamsTopologyEndpointAutoConfiguration.class in our @SpringBootApplication annotation.
`@SpringBootApplication(exclude= {KafkaStreamsTopologyEndpointAutoConfiguration.class})`

**Comment by sobychacko** [comment]
Reopening the issue as this is still not resolved.

**Comment by sobychacko** [comment]
This is addressed through this change on 3.0.x: https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/commit/f0bd9f00c2e7132354fff1acf5d5d3e720e07bf8

Same change has been applied to master as well: https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/commit/0a0d3a105708a5dfcd11c7711c3abfc9fea76d2f

@yonigibbs The typo is now fixed. 

The next release of Horsham.SR9 (3.0.9) will contain this fix.
Thank you! Please feel free to re-open the issue, if anyone still seeing any issues.
